proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,309 @@
+	  mtc.prompt_for_task("question-generation", documents=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ['What is the capital of Germany?']
+  ```
+
+- MultiTaskComponent supports selecting a particular current task (e.g. question-generation) and then subsequently
+  using the selected task until user changes the current task. For example:
+
+  ``` python
+	  qa = mtc.task("question-generation")
+      qa(documents=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ['What is the capital of Germany?']
+  ```
+
+- The addition of new tasks is supported by the `add_task` method. For example:
+
+  ``` python
+      from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      mtc.add_task(TaskTemplate(name="sentiment-analysis",
+                          prompt_text="Please give a sentiment for this context. Answer with positive, negative or neutral. Context: $documents; Answer:",
+                          prompt_params=["documents"]))
+      mtc.supported_task_names()
+
+      ----------------------------------------------------------------------------
+      ['question-answering',
+       'question-generation',
+       'conditioned-question-generation',
+       'summarization',
+       'question-answering-check',
+       'sentiment-analysis']
+  ```
+
+- Task parameters verification
+
+  All task parameters will be verified to match the task template and the corresponding object parameters will be
+  checked for type and value. For example:
+
+  ``` python
+      from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      mtc.prompt_for_task("question-generation", some_unknown_param=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ValueError                                Traceback (most recent call last)
+      <ipython-input-16-369cca52e960> in <module>
+            1 # tasks parameters are checked
+      ----> 2 sa(some_param=[Document("Berlin is the capital of Germany.")])
+
+      2 frames
+      /usr/local/lib/python3.8/dist-packages/haystack/nodes/llm/multi_task.py in __call__(self, *args, **kwargs)
+           34         if set(template_dict.keys()) != set(self.prompt_params):
+           35             available_params = set(list(template_dict.keys()) + list(set(kwargs.keys())))
+      ---> 36             raise ValueError(f"Expected prompt params {self.prompt_params} but got {list(available_params)}")
+           37
+           38         template_dict["prompt_template"] = self.prompt_text
+
+      ValueError: Expected prompt params ['documents'] but got ['some_unknown_param']
+  ```
+
+- MultiTaskPipelines
```

Comment:
I haven't checked the implementation yet, but I wonder what makes `MultiTaskPipeline` special wrt to a basic Pipeline. Should we rather improve Pipeline to support a `MultiTaskComponent`, or extend `Task` to make the output of this node compatible with existing Haystack pipelines?

If it's matter of adding a `task` parameter to `run()`, I see it doable without having to create a separate `Pipeline` class. Not going to be neat, but better than creating a custom `Pipeline` that we will have to refactor later along with the basic one imho.

Reply:
I am not sure either what's the best option here. So far, I have managed to pack all parameters into meta parameters in MultiTaskPipeline and then unpack them in MultiTaskComponent and use them for invocation. Can we do that until we have Haystack 2.0 pipelines?

Reply:
@ZanSara addressed your pipeline suggestions in the latest commit. Resolving this conversation